torch/jit/__init__.py

```diff
@@ -621,29 +622,88 @@ def env(key):
 
 class CompilationUnit(object):
     def __init__(self, lang=None, optimize=True):
-        self.cu = torch._C.CompilationUnit()
+        self.module = torch._C.ScriptModule(optimize)
         if lang is not None:
             self.define(lang, frame_id=3)
-        self.execution_engines = {}
         self.optimize = optimize
 
     def define(self, lang, rcb=None, frame_id=2):
         if not rcb:
             rcb = createResolutionCallback(frame_id)
-        self.cu.define(lang, rcb)
+        self.module._define(lang, rcb, False)
 
     def __getattr__(self, attr):
-        if attr not in self.execution_engines:
-            graph = self.cu.get_graph(attr)
-            self.execution_engines[attr] = torch._C.GraphExecutor(graph, self.optimize)
-        return self.execution_engines[attr]
+        return self.module._get_method(attr)
 
 
 def script(fn):
     rcb = createResolutionCallback()
-    ast = torch.jit.frontend.get_jit_ast(fn)
+    ast = get_jit_ast(fn)
     graph = _jit_script_compile(ast, rcb)
     return torch._C.GraphExecutor(graph, True)
 
+
+ScriptMethodStub = namedtuple('ScriptMethodStub', ('resolution_callback', 'ast'))
+
+
+def script_method(fn):
+    return ScriptMethodStub(createResolutionCallback(), get_jit_ast(fn))
+
+
+class ScriptMeta(type(torch._C.ScriptModule)):
+    # this has to inherit from pybind11's metaclass otherwise we get
+    # issues because ScriptModule inherits from torch._C.ScriptModule,
+    # a pybind11 type
+    def __init__(cls, name, bases, attrs):
+        # find all the script methods
+        methods = []
+        for k, v in sorted(attrs.items()):
+            if isinstance(v, ScriptMethodStub):
+                delattr(cls, k)
+                methods.append(v)
+        if len(methods) > 0 and '__init__' in attrs:
+            # after the user's __init__ register all the script methods
+            # with the module
+            original_init = cls.__init__
+
+            def init_then_register(self, *args, **kwargs):
+                original_init(self, *args, **kwargs)
+                for m in methods:
+                    self._create_method(m.ast, m.resolution_callback)
+
+            cls.__init__ = init_then_register
+        return super(ScriptMeta, cls).__init__(name, bases, attrs)
+
+
+class ScriptModule(with_metaclass(ScriptMeta, torch._C.ScriptModule)):
```

Comment:
Can we have a doc describing the general metaclass strategy here? Metaclasses are pretty unusual Python so a doc will help anyone who isn't a metaclass expert to work around this code.

Reply:
@zdevito and you complain when I submit patches that use metaclasses

Reply:
I am open to ideas where I don't need a metaclass, but I don't know how to make it work otherwise :-P We need to register and specialize all script methods _after_ `__init__` has run in a subclass, and I didn't want to force the user to call `self.after_init()`, because the error messages when you forget it would be bad. Making script method initialization lazy would also have bad properties, since it can produce user-facing error message. I will document this better.

Reply:
I thought the metaclass was pretty clear once I actually read the metaclass implementation. The confusion was more organizational (things put where and why)?

Reply:
Oh I was just joking. I’m ok with this solution